src/v/cluster/cloud_metadata/offsets_lookup_batcher.cc

```diff
@@ -0,0 +1,148 @@
+            })
+          .then(&rpc::get_ctx_data<offsets_lookup_reply>);
+    if (reply.has_error()) {
+        vlog(
+          clusterlog.error,
+          "Error sending offsets request for {} NTPs to node {}: {}",
+          num_requested_ntps,
+          empty_reply.node_id,
+          reply.error());
+        co_return empty_reply;
+    }
+    co_return std::move(reply.value());
+}
+
+ss::future<> offsets_lookup_batcher::run_lookups(
+  absl::btree_set<model::ntp> ntps_to_lookup, retry_chain_node& retry_node) {
+    // Keep sending out lookup requests until a non-retriable error is hit, or
+    // until there are no more NTPs to retry.
+    while (!ntps_to_lookup.empty()) {
+        auto retry_permit = retry_node.retry();
+        if (!retry_permit.is_allowed) {
+            vlog(
+              clusterlog.error,
+              "Timed out running lookups with {} NTPs to lookup",
+              ntps_to_lookup.size());
+            throw ss::timed_out_error();
+        }
+        vlog(clusterlog.debug, "Looking up {} NTPs", ntps_to_lookup.size());
+        std::vector<ss::future<offsets_lookup_reply>> pending_replies;
+        absl::btree_map<model::node_id, offsets_lookup_request>
+          pending_requests_per_node;
+
+        // Copy the NTPs to send. As we process duplicates, we may end up
+        // removing from our lookup set.
+        fragmented_vector<model::ntp> ntps(
+          ntps_to_lookup.begin(), ntps_to_lookup.end());
+
+        // Batch up all our lookup requests per node.
+        for (auto& ntp : ntps) {
+            if (_offset_by_ntp.contains(ntp)) {
+                ntps_to_lookup.erase(ntp);
+                continue;
+            }
+            auto leader_id = _leaders_table.get_leader(ntp).value_or(
+              model::node_id{});
+            if (leader_id == model::node_id{}) {
+                // Perhaps no leader? Retry the ntp in the next iteration.
+                continue;
+            }
+            auto& node_req = pending_requests_per_node[leader_id];
+            node_req.node_id = leader_id;
+            node_req.ntps.emplace_back(std::move(ntp));
+
+            // Once we hit the target per node, send out the request.
+            if (node_req.ntps.size() == _batch_size) {
+                pending_replies.emplace_back(send_request(std::move(node_req)));
+                pending_requests_per_node.erase(leader_id);
+            }
+        }
+        // Send all remaining batched requests, regardless of size.
+        for (auto& [node_id, node_req] : pending_requests_per_node) {
+            pending_replies.emplace_back(send_request(std::move(node_req)));
+        }
+
+        // Materialize the replies from the futures.
+        auto per_node_replies = co_await ss::when_all(
+          pending_replies.begin(), pending_replies.end());
+        for (auto& r : per_node_replies) {
+            vassert(r.available(), "waited for future but not available");
+            auto reply = r.get();
```

Comment:
a bit surprised to see given how much work is happening in this function that if `r` holds an exception that we abandon all the work that's been performed. is that true / intended?

Reply:
Nice catch, not intentional! Done